### DISCUSSION TITLE
Compared to Ansible

### DISCUSSION THREAD

**Discussion by n8th3n** [pyinfra-dev/pyinfra#976] [Q&A]
Hi
The feature set appears to overlap with Ansible, so how does it compare? Why use pyinfra over Ansible?

**Comment by sfermigier** [comment]
I personally use Pyinfra rather than Ansible because I don't think it's a good idea to use YAML as a programming language.

**Comment by Fizzadar** [comment]
Ansible is one of the reasons I stated pyinfra in the first place! Some things that I wanted to solve (note this is coming from ansible 2.1):

- debugging, pyinfra executes shell commands just like a human would and if there’s an error you can see the output right away. This was my greatest frustration with ansible
- speed, at the time ansible didn’t have control persist or other things and was crazy slow, pretty sure this has changed now, I’ve been meaning to redo the perf comparison but it’s pretty low on my priority currently!
- using python not yaml, as above, I really hate configuration markup twisted into programming code, or some horrible attempt at that. Terraform and helm have the same problem
- knowing what’s happening - I always designed pyinfra around the idea of “execute shell command, parse output, execute another shell command” because it’s the same as a human would when setting stuff up by hand. Because of this it’s possible to print out all the commands and output in real-time, there’s no abstraction inside python code executing on the remote side

That’s all I can think of- this is all way back from 2015 when I first started pyinfra! I know both ansible and pyinfra have changed a lot since then, but the general philosophy of how pyinfra works to this day is as above.

**Reply by tidalvirus** [reply]
This would be awesome in an FAQ - I only came across pyinfra this week, and was wondering about the 'why' of it :)

**Comment by vojta001** [comment]
As others have written, driving the deploys with real code is super strong. It isn't just whether you write loops as `for v in arr:` or using YAML constructs.

I manage ethernet switches using pyinfra, and the process looks like

1. Connect to our internal database to find out the **desired configuration** of the devices (perfect task for `inventory.py`)
2. Collect the **current configuration** (just load some pyinfra's _facts_)
3. Compute the diff and build the needed commands (our "business logic" simply written in python inside the _deploy_ file)
4. Execute the commands (spill out some shell commands of some custom _operations_)

Good luck doing 1. and 3. with ansible.

On the other hand, ansible's ecosystem is much larger and more mature, especially the modules are better polished. If you want execute rather standard operations on rather standard targets, ansible may save you some time in the short term.
